cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go

```diff
@@ -398,6 +426,51 @@ func (c *client) DeleteMachineObjectsInNamespace(namespace string) error {
 	return nil
 }
 
+func (c *client) ForceDeleteMachine(namespace, name string) error {
+	machine, err := c.clientSet.ClusterV1alpha1().Machines(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "error getting Machine %s/%s", namespace, name)
+	}
+	machine.SetFinalizers([]string{})
+	if _, err := c.clientSet.ClusterV1alpha1().Machines(namespace).Update(machine); err != nil {
+		return errors.Wrapf(err, "error removing finalizer for Machine %s/%s", namespace, name)
+	}
+	if err := c.clientSet.ClusterV1alpha1().Machines(namespace).Delete(name, newDeleteOptions()); err != nil {
+		return errors.Wrapf(err, "error deleting Machine %s/%s", namespace, name)
+	}
+	return nil
+}
+
+func (c *client) ForceDeleteMachineSet(namespace, name string) error {
+	ms, err := c.clientSet.ClusterV1alpha1().MachineSets(namespace).Get(name, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "error getting MachineSet %s/%s", namespace, name)
+	}
+	ms.SetFinalizers([]string{})
+	if _, err := c.clientSet.ClusterV1alpha1().MachineSets(namespace).Update(ms); err != nil {
+		return errors.Wrapf(err, "error removing finalizer for MachineSet %s/%s", namespace, name)
+	}
+	if err := c.clientSet.ClusterV1alpha1().MachineSets(namespace).Delete(name, newDeleteOptions()); err != nil {
```

Comment:
What finalizer are we deleting before adding the foreground deletion finalizer?

Reply:
We are deleting the finalizer for the controller, otherwise the delete stays pending until a reconciliation is performed. For example, we set the finalizer for Clusters here: https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/controller/cluster/controller.go#L95

Reply:
ok cool, makes sense 👍